File: apps/api/app/schemas/evaluation.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class EvaluationResultResponse(BaseModel):
    id: uuid.UUID
    agent_run_id: uuid.UUID | None
    case_id: str
    category: str
    question: str
    expected_answer: str | None
    answer: str | None
    passed: bool
    abstained: bool
    should_abstain: bool
    retrieval_recall: float
    retrieval_precision: float
    citation_accuracy: float
    faithfulness: float
    answer_relevance: float
    hallucinated: bool
    latency_ms: float
    input_tokens: int
    output_tokens: int
    cost_usd: float

    model_config = {"from_attributes": True}


class RegressionResponse(BaseModel):
    metric: str
    baseline: float
    current: float
    delta: float


class EvaluationRunResponse(BaseModel):
    id: uuid.UUID
    dataset_version: str
    status: str
    error_message: str | None
    total_cases: int
    passed_cases: int
    failed_cases: int
    faithfulness: float | None
    citation_accuracy: float | None
    retrieval_recall: float | None
    retrieval_precision: float | None
    hallucination_rate: float | None
    answer_relevance: float | None
    avg_latency_ms: float | None
    avg_input_tokens: float | None
    avg_output_tokens: float | None
    avg_cost_usd: float | None
    baseline_run_id: uuid.UUID | None
    regressions: list[RegressionResponse]
    created_at: datetime

    model_config = {"from_attributes": True}


class EvaluationRunDetailResponse(EvaluationRunResponse):
    results: list[EvaluationResultResponse]


class EvaluationRunListResponse(BaseModel):
    evaluation_runs: list[EvaluationRunResponse]

File: apps/web/types/evaluation.ts
export interface EvaluationResult {
  id: string;
  agent_run_id: string | null;
  case_id: string;
  category: string;
  question: string;
  expected_answer: string | null;
  answer: string | null;
  passed: boolean;
  abstained: boolean;
  should_abstain: boolean;
  retrieval_recall: number;
  retrieval_precision: number;
  citation_accuracy: number;
  faithfulness: number;
  answer_relevance: number;
  hallucinated: boolean;
  latency_ms: number;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
}

export interface EvaluationRegression {
  metric: string;
  baseline: number;
  current: number;
  delta: number;
}

export interface EvaluationRun {
  id: string;
  dataset_version: string;
  status: "running" | "completed" | "failed";
  error_message: string | null;
  total_cases: number;
  passed_cases: number;
  failed_cases: number;
  faithfulness: number | null;
  citation_accuracy: number | null;
  retrieval_recall: number | null;
  retrieval_precision: number | null;
  hallucination_rate: number | null;
  answer_relevance: number | null;
  avg_latency_ms: number | null;
  avg_input_tokens: number | null;
  avg_output_tokens: number | null;
  avg_cost_usd: number | null;
  baseline_run_id: string | null;
  regressions: EvaluationRegression[];
  created_at: string;
}

export interface EvaluationRunDetail extends EvaluationRun {
  results: EvaluationResult[];
}
